src/scripts/validate_synapse_records/validate_synapse_records.py

```diff
@@ -0,0 +1,172 @@
+
+The records must use the self-referencing format in their metadata.json file.
+Any files in the record which do not have a `jsonSchema` property specified in
+their fileInfo object in metadata.json will not be validated.
+
+The validation results are written to `validation_errors.json`.
+"""
+
+import json
+import zipfile
+import argparse
+import boto3
+import jsonschema
+import requests
+import synapseclient
+
+def read_args():
+    parser = argparse.ArgumentParser(
+            description="Validate records in a Synapse file view against their JSON Schema.")
+    parser.add_argument("--file-view",
+                        required=True,
+                        help=("The Synapse ID of a file view containing "
+                              "files to be submitted."))
+    parser.add_argument("--query",
+                        default="select * from {source_table}",
+                        help=("Optional. An f-string formatted query which filters the "
+                              "file view. Use {source_table} in the FROM clause. "
+                              "If this argument is not specified, all files in "
+                              "the --file-view will be submitted to the "
+                              "--glue-workflow."))
+    parser.add_argument("--ssm-parameter",
+                        help=("Optional. The name of the SSM parameter containing "
+                              "the Synapse personal access token. "
+                              "If not provided, cached credentials are used"))
+    parser.add_argument("--profile",
+                        help=("Optional. The AWS profile to use. Uses the default "
+                              "profile if not specified."))
+    args = parser.parse_args()
+    return args
+
+def get_synapse_client(ssm_parameter=None, aws_session=None):
+    """
+    Return an authenticated Synapse client.
+
+    Args:
+        ssm_parameter (str): Name of the SSM parameter containing the
+            BridgeDownstream Synapse password.
+        aws_session (boto3.session.Session)
+
+    Returns:
+        synapseclient.Synapse
+    """
+    if ssm_parameter is not None:
+        ssm_client = aws_session.client("ssm")
+        token = ssm_client.get_parameter(
+            Name=ssm_parameter,
+            WithDecryption=True)
+        syn = synapseclient.Synapse()
+        syn.login(authToken=token["Parameter"]["Value"])
+    else: # try cached credentials
+        syn = synapseclient.login()
+    return syn
+
+def get_schemas(synapse_files: list):
+    """
+    Get schemas for each JSON file in each Synapse file in a list of Synapse files.
+
+    Args:
+        synapse_files (list): a list of synapseclient.entity.File objects.
+
+    Returns:
+        all_schemas (dict): A dictionary of schemas
+            {
+                "synapseId": {
+                    "filename": "schema_url",
+                    ...
+                },
+            }
+    """
+    all_schemas = {}
```

Comment:
Nit: look into collections.defaultdict.

Reply:
I dislike the idea of importing an additional library to cover such a simple use case.